runtime/jcl/common/jcldefine.c

```diff
@@ -160,7 +160,33 @@ defineClassCommon(JNIEnv *env, jobject classLoaderObject,
 			/* Bad, we have already defined this class - fail */
 			omrthread_monitor_exit(vm->classTableMutex);
 			if (J9_ARE_NO_BITS_SET(*options, J9_FINDCLASS_FLAG_NAME_IS_INVALID)) {
-				vmFuncs->setCurrentExceptionNLSWithArgs(currentThread, J9NLS_JCL_DUPLICATE_CLASS_DEFINITION, J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR, utf8Length, utf8Name);
+				/* TODO: This path is taken if Classloader.findClass is called on a persisted class.
+				 * Once class objects are persisted, reaching this point is an error.
+				 */
+				if (!IS_RESTORE_RUN(vm)) {
+					vmFuncs->setCurrentExceptionNLSWithArgs(
+						currentThread,
+						J9NLS_JCL_DUPLICATE_CLASS_DEFINITION,
+						J9VMCONSTANTPOOL_JAVALANGLINKAGEERROR,
+						utf8Length,
+						utf8Name);
+#if defined(J9VM_OPT_SNAPSHOTS)
+				} else {
+					clazz = vmFuncs->hashClassTableAt(classLoader, utf8Name, utf8Length);
+
+					if (!vmFuncs->loadWarmClassFromSnapshot(currentThread, classLoader, clazz)) {
+						clazz = NULL;
+					}
+
+					clazz = vmFuncs->initializeSnapshotClassObject(vm, classLoader, clazz);
+					if (NULL != protectionDomain) {
+						J9VMJAVALANGCLASS_SET_PROTECTIONDOMAIN(
+							currentThread,
+							clazz->classObject,
+							J9_JNI_UNWRAP_REFERENCE(protectionDomain));
+					}
+#endif /* defined(J9VM_OPT_SNAPSHOTS) */
```

Comment:
I think the previous version of this was better (with the test of `IS_RESTORE_RUN()` within the `J9VM_OPT_SNAPSHOTS` block). This way may leave readers wondering if nothing should happen if `IS_RESTORE_RUN()`, when `J9VM_OPT_SNAPSHOTS` is not defined.